### PR TITLE
Docs

### PR DESCRIPTION
**Pull request by pat-alt**
Some early work on regression docs (-> blog post). In regions of the domain that were not seen during training, predictive uncertainty should be really high (no?). See, for example, my results for a Bayesian Neural Network lifted from [LaplaceRedux.jl](https://github.com/pat-alt/LaplaceRedux.jl#regression):

image
 
When I run a similar experiment using CP (here split conformal for XGBoost), results look very different:
 
image
 
Considering how CP works, I don’t find this result too surprising, it’s just not great and I’m now a bit confused as to how exactly the coverage guarantees work.
 
The chart is produced in this [notebook](https://github.com/pat-alt/ConformalPrediction.jl/blob/docs/docs/src/regression.qmd), where I’ve been trying to replicate this [tutorial](https://mapie.readthedocs.io/en/latest/examples_regression/4-tutorials/plot_main-tutorial-regression.html#sphx-glr-download-examples-regression-4-tutorials-plot-main-tutorial-regression-py) from MAPIE.  

It would be useful to cross-check the results against the MAPIE results (notebook stored [here](https://github.com/pat-alt/ConformalPrediction.jl/blob/docs/dev/plot_main-tutorial-regression.ipynb)). In there they don't really look at values of `x` outside of the training range at the prediction stage, but it should be straight-forward to make adjustments to the code in the notebook. (**EDIT**: done below)

### CODE REVIEW

**Comment by pat-alt** [comment]
I've now run the MAPIE tutorial with gradient boosting instead of polynomial reg and the results are the same (so this is not an implementation issue):

image

Quite puzzled now: looking at the chart, test coverage left and right of the training points is essentially zero. I'm probably just misunderstanding the coverage guarantee.

**Comment by pat-alt** [comment]
When using polynomial regression as opposed to gradient boosting, the results look better (in some cases). But this is driven by the fact that separate fits for subsets of the data (CV, Jackknife, ...) reflect the underlying stochasticity of the data.  

image

Again, I suppose exchangeability is violated: I draw training data from $\mathcal{X}\sim\text{Unif}(-3,3)$ and then map to $\mathcal{Y}$. Test data left and right of the training data is not drawn from $\mathcal{X}$, so we have covariate shift (will be interesting to look at [Section 4.5]((https://arxiv.org/pdf/2107.07511.pdf)) in this context). Is that right @aangelopoulos? Thanks in advance and apologies if trivial (still fairly new territory for me and trying to wrap my head around things).

**Comment by aangelopoulos** [comment]
Hey @pat-alt !  Happy to help!

I think you have the right idea --- the example that you've given comes from an extreme covariate shift.
If you run covariate shift conformal using Section 4.5 of the gentle intro, what will happen is that your intervals will explode outside the training data, since the likelihood ratio gets very large.

Without using the covariate shift technique, conformal will only give you the marginal guarantee if the new data is i.i.d. with your old data... so the regions with no training data don't really matter, since they occur with very low probability.

My suggestion would be this.  If you like the way the Bayesian NN performs in this setting, use it as your heuristic notion of uncertainty. Then calibrate it with conformal. Best of both worlds!

**Comment by pat-alt** [comment]
> Hey @pat-alt ! Happy to help!
> 
> I think you have the right idea --- the example that you've given comes from an extreme covariate shift. If you run covariate shift conformal using Section 4.5 of the gentle intro, what will happen is that your intervals will explode outside the training data, since the likelihood ratio gets very large.
> 
> Without using the covariate shift technique, conformal will only give you the marginal guarantee if the new data is i.i.d. with your old data... so the regions with no training data don't really matter, since they occur with very low probability.
> 
> My suggestion would be this. If you like the way the Bayesian NN performs in this setting, use it as your heuristic notion of uncertainty. Then calibrate it with conformal. Best of both worlds!

Got it - thank you very much! 

I'm gradually working my way through your tutorial (which is fantastic by the way!). Both Conformalized Bayes #31 and covariate shift conformal #39 are on the to-do list 😄 

Thanks again for your help - much appreciated!

**Comment by aangelopoulos** [comment]
Thanks Pat!!  This package that you are building is great!!
